### PR TITLE
Proposed change - make toasts usable in tests

### PR DESCRIPTION
**Pull request by kcappieg**
Bringing file inline with lwc-recipe:
https://github.com/salesforce/sfdx-lwc-jest/blob/master/src/lightning-stubs/platformShowToastEvent/platformShowToastEvent.js

### CODE REVIEW

File: src/lightning-stubs/platformShowToastEvent/platformShowToastEvent.js [hunk 1]
```diff
@@ -4,14 +4,15 @@
  * SPDX-License-Identifier: MIT
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
  */
-const ShowToastEventName = 'lightning__showtoast';
+export const ShowToastEventName = 'lightning__showtoast';
 
 export class ShowToastEvent extends CustomEvent {
-    constructor() {
+    constructor(toast) {
```
Comment (pmdartus) [inline src/lightning-stubs/platformShowToastEvent/platformShowToastEvent.js]:
```suggestion
    constructor(detail) {
```

Comment (kcappieg) [reply src/lightning-stubs/platformShowToastEvent/platformShowToastEvent.js]: Note to reviewers: If this line changes to match the suggestion, so will line 15

**Comment by sfdciuie** [comment]
@pmdartus @kcappieg  The internal details of that event are not meant to be ever processed by anyone but the core platform (this doesn't work at all off-core), personally I don't think it's a good idea to expose anything beyond a dumb stub, because any test that relies on this will likely fail at one point in the future as it would depend on internals of something that is not an api.

**Comment by muenzpraeger** [comment]
We had to mock out our own implementation for LWC Recipes [here](https://github.com/trailheadapps/lwc-recipes/blob/master/force-app/test/jest-mocks/lightning/platformShowToastEvent.js) too, because the current stub doesn't provide the needed information.

**Comment by sfdciuie** [comment]
@muenzpraeger That doesn't change the fact that the test then would rely on internal implementation details instead of actual api, and if something is meant to be private, then introspecting it isn't the right thing to do.

**Comment by muenzpraeger** [comment]
@sfdciuie So what’s the solution then if want to test if a toast sends the correct detail information, which I have for example constructed via code?

**Comment by kcappieg** [comment]
@sfdciuie The main thing I want to test is that toasts are being fired. I have requirements for my components that notifications must be displayed to users in various circumstances. I understand your argument about reaching into internals, but it doesn't seem like an internal detail to at least be able to verify that a toast is fired via exposing the event name variable with which we can attach event listeners.

I think it's reasonable to modify my PR such that the only change is exporting the `ShowToastEventName` variable and exclude the `details` property change.

**Comment by kcappieg** [comment]
> @sfdciuie So what’s the solution then if want to test if a toast sends the correct detail information, which I have for example constructed via code?

@muenzpraeger workaround could be to store the details in a "test-only" property. Maybe even create a utility around sending toasts that you can spy on in other components so you can isolate the code that actually fires the toast and provide a good way to verify what's being passed to your toasts.

**Comment by muenzpraeger** [comment]
@kcappieg I see this at something that should be resolved at the (test) framework level, so that not everyone has to invent their own solution.

**Comment by pmdartus** [comment]
Now that I am thinking about it, you probably don't need to expose the event name out of the module for  now. You can create an event in your test file and extract the event `name` property from there to attach an handler, for example:

```js
const toastEventName = new ShowToastEvent().name;

test('component should fire a toast event', () => {
  document.addEventListener(toastEventName, () => {
    console.log('toast event fired');
  });

  // Render some component and let it trigger the toast event.
});
```

That being said, until we add the `detail` to the stub you will not be able to observe from your test what payload was passed to the toast.

**Comment by kcappieg** [comment]
I suppose that's true, didn't think of that. But personally that seems less intuitive than just exporting the event name. Why should we have to construct a new object we're going to throw away when the name is identical to what is being exported in my PR.

I've been thinking about @sfdciuie 's comment about the internals: we are not _really_ reaching into internals of how the toast works if only providing a way to access what was passed into it. Theoretically the production version of that event could mangle that object as much as it wants for its event object that it dispatches, but in the test setting the toast is not being processed so why can't the toast configuration just be passed up as the detail?

Since this is specifically a mock version of this event, it seems to me that it can behave differently to facilitate testing.

**Comment by sfdciuie** [comment]
@kcappieg In my opinion the issue with mocks isn't that one is modifying the internals of an existing implementation (as you said as well it's just a mock, actual production code isn't being changed), but rather that one is assuming how a component works internally, it's ok if APIs are stubbed, but it's not ok to assume for internal mechanisms to continue existing - event name may change, the event may not bubble any more, but the functionality will continue to work. If one wants to test whether the notification worked, then it needs to be done on the platform via a a functional/integration test. If one wants to test that the details associated with a notification are correct, I would suggest instead to create a function that encapsulates notification emittance, so that you now have added flexibility in how notifications are emitted and full control over what's testable as it's your own API. 

It's also fine to add a direct property to the mock event that would return you exactly what you passed into this, but I don't think it's a good idea to add this as a `detail` of the event as that starts to mimic non-public API of the event.

**Comment by kcappieg** [comment]
@sfdciuie as far as name changing is concerned, the name is exported from the test file, so user code has no dependency on a hard-coded string.

Reread @pmdartus 's code snippet and noticed that they're attaching an event listener to `document`, which I can understand your argument against. My tests always attach an event listener directly to the component firing off the event which would be resilient to changing bubbling status in most circumstances.

Ultimately my goal here is to make your testing framework more pleasant to use since right now our tests have required a patchwork of hacks and workarounds for testing gaps. I hear you on the public API vs internals argument, but you're effectively saying I can't unit test toast notifications at all even though all I'm trying to verify is the public API (i.e. it was fired with these parameters at this time from this component). I don't see why we can't get a mock, even with some restrictive assumptions like all propagation flags set to `false`.

The last thing I'll say about the accessible `detail`: to me it seems less like mimicking production code and more like adding a "spy" to the toast event process. I could put a proxy in front of the toast notification, sure, but that seems like overkill. A compromise could be that the test version of the event attaches the passed detail object as a separate test-specific property to the event:

```Javascript
super(ShowToastEventName, {
  composed: true,
  cancelable: true,
  bubbles: true
});

this.test_ToastConfigSpy = detail;
```

But speaking of overkill, if you guys are dead set on keeping this file the way it is then go ahead and close this PR. I will work around. It seemed like a typo that it was missing, especially since some of your "recipes" include test files _with_ the evt name exported, so I actually thought this was just a quick win.

**Comment by trevor-bliss** [comment]
Apologies @kcappieg, but I think you should continue to use your own stub for now until we can reach agreement on the best way to solve this at the framework level.

**Pull request closed without merging**